File: notes/windows/Amadeuz/LocalStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Amadeuz;

/// <summary>
/// Persists folders and notes to %APPDATA%\amadeuz\data.json.
/// Same JSON schema as the macOS and server.
/// Writes are atomic (temp file + rename) so a crash mid-write can't corrupt the file.
/// </summary>
public sealed class LocalStore
{
    private readonly string _filePath;

    public LocalStore()
    {
        var dir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "amadeuz");
        Directory.CreateDirectory(dir);
        _filePath = Path.Combine(dir, "data.json");
    }

    public (List<Folder> folders, List<Note> notes) Load()
    {
        try
        {
            if (!File.Exists(_filePath)) return (new(), new());
            var doc = JsonSerializer.Deserialize<LocalData>(File.ReadAllText(_filePath));
            return (doc?.Folders ?? new(), doc?.Notes ?? new());
        }
        catch { return (new(), new()); }
    }

    public void Save(IEnumerable<Folder> folders, IEnumerable<Note> notes)
    {
        try
        {
            var tmp = _filePath + ".tmp";
            File.WriteAllText(tmp, JsonSerializer.Serialize(
                new LocalData { Folders = new(folders), Notes = new(notes) }));
            File.Move(tmp, _filePath, overwrite: true);
        }
        catch { }
    }

    private sealed class LocalData
    {
        [JsonPropertyName("folders")] public List<Folder> Folders { get; init; } = new();
        [JsonPropertyName("notes")]   public List<Note>   Notes   { get; init; } = new();
    }
}

File: notes/android/app/src/main/java/com/amadeuz/notes/LocalStore.kt
package com.amadeuz.notes

import android.content.Context
import com.google.gson.Gson
import java.io.File

private data class LocalData(val folders: List<Folder>, val notes: List<Note>)

class LocalStore(context: Context) {
    private val file = File(context.filesDir, "data.json")
    private val gson = Gson()

    fun load(): Pair<List<Folder>, List<Note>> {
        if (!file.exists()) return Pair(emptyList(), emptyList())
        return try {
            val data = gson.fromJson(file.readText(), LocalData::class.java)
            Pair(data?.folders ?: emptyList(), data?.notes ?: emptyList())
        } catch (_: Exception) {
            Pair(emptyList(), emptyList())
        }
    }

    fun save(folders: List<Folder>, notes: List<Note>) {
        try {
            file.writeText(gson.toJson(LocalData(folders, notes)))
        } catch (_: Exception) {}
    }
}
